### PR TITLE
Market Plugin: Ability to mark as sold

### PR DESCRIPTION
**Pull request by tanndlin**

### CODE REVIEW

File: src/app/plugins/marketplace.plugin.ts [hunk 1]
```diff
@@ -98,6 +99,15 @@ export class MarketPlacePlugin extends Plugin {
...
           item += '\n' + user; //adds user to end of listing.
+
+          //Check if sold
+          const hasTargetReaction = msg.reactions.find(
+            (r) => r.emoji.name === this._TARGET_REACTION
+          );
+          if (hasTargetReaction) {
+            item += '\t SOLD';
+          }
+
```
Comment (c650) [inline src/app/plugins/marketplace.plugin.ts:110]: why include the listing at all if sold? and maybe you should check that the reaction is from the message.author

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:110]: People were complaining that they'd want to see sold listings. The author can still delete the message if they'd like

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:110]: I'll do the other part in a little bit

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:110]: Do you know why `console.log(hasTargetReaction.users)` prints an empty collection?

Comment (c650) [reply src/app/plugins/marketplace.plugin.ts:110]: probs some async thing

File: src/app/plugins/marketplace.plugin.ts [hunk 2]
```diff
@@ -104,13 +105,19 @@ export class MarketPlacePlugin extends Plugin {
           const hasTargetReaction = msg.reactions.find(
             (r) => r.emoji.name === this._TARGET_REACTION
           );
-          if (hasTargetReaction) {
-            item += '\t SOLD';
+
+          if (hasTargetReaction === null) {
```
Comment (c650) [inline src/app/plugins/marketplace.plugin.ts:109]: !hasTargetReaction

File: src/app/plugins/marketplace.plugin.ts [hunk 3]
```diff
@@ -93,14 +94,30 @@ export class MarketPlacePlugin extends Plugin {
...
+            itemsForSale.push(item);
+            return;
+          }
+
+          hasTargetReaction.fetchUsers().then((users) => {
+            if (users.has(msg.author.id)) {
+              item += '\t SOLD';
+            }
+            itemsForSale.push(item);
+          });
```
Comment (c650) [inline src/app/plugins/marketplace.plugin.ts:119]: do we ever await this?

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:119]: You mean i shouldn't await it?

Comment (c650) [reply src/app/plugins/marketplace.plugin.ts:119]: you have to await it because you need the push to happen before you build the message right?

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:119]: Correct. I used then, doesn't that work?

Comment (c650) [reply src/app/plugins/marketplace.plugin.ts:119]: `then` unwraps the promise, does some stuff, and then returns a new promise. So it doesn't necessarily work afaik.

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:119]: Oh, ok. I was having trouble using await inside of a forEach, I'll make a queue or something instead

**Comment by tanndlin** [comment]
Most of that was moving bulks of code into their own functions

File: src/app/plugins/marketplace.plugin.ts [hunk 4]
```diff
@@ -77,32 +78,62 @@ export class MarketPlacePlugin extends Plugin {
...
       const batch = await message.channel.fetchMessages(config);
+      //Make sure there are messages
+      if (!batch.array().length) {
+        continue;
+      }
+      last_id = batch.last().id;
+
+      const calls = batch.filter((msg) => this._startsWithPrefix(msg)); //Filter out non !market adds
+      const parsed = calls.map((msg) => this._resolveToListing(msg)); //Turn them into listings
+      parsed.forEach((e) => buffer.push(e));
```
Comment (c650) [inline src/app/plugins/marketplace.plugin.ts:94]: does `parsed.push(...parsed)` work ?

File: src/app/plugins/marketplace.plugin.ts [hunk 5]
```diff
@@ -77,32 +78,62 @@ export class MarketPlacePlugin extends Plugin {
...
-        let [, item] = content.split('add');
-        if (item?.length) {
-          const user = msg.author;
-          item += '\n' + user; //adds user to end of listing.
-          itemsForSale.push(item);
-        }
-        last_id = msg.id; //last_id will end up as the last message's id
+    await Promise.all(buffer).then((items) => {
+      items.forEach((msg) => {
+        itemsForSale.push(msg);
```
Comment (c650) [inline src/app/plugins/marketplace.plugin.ts:99]: saa

File: src/app/plugins/marketplace.plugin.ts [hunk 6]
```diff
@@ -77,32 +78,62 @@ export class MarketPlacePlugin extends Plugin {
...
+    let [, item] = content.split('add');
+
+    if (!item?.length) {
+      return '';
+      /*The messages are already filtered before this function is called
+      So this should theoretically never be true*/
+    }
+
+    const user = msg.author;
+    item += '\n' + user; //adds user to end of listing.
```
Comment (c650) [inline src/app/plugins/marketplace.plugin.ts:116]: do you have prettier installed? `//<space>comment`

Comment (tanndlin) [reply src/app/plugins/marketplace.plugin.ts:116]: Yeah, i just formatted with prettier and nothing changed

**Review by c650 — APPROVED** [review]

**Pull request merged**